hpx/lcos/future.hpp

```diff
@@ -1041,53 +950,20 @@ namespace hpx { namespace lcos
         using base_type::has_exception;
 
         template <typename F>
-        typename util::lazy_enable_if<
-            !hpx::traits::is_launch_policy<
-                typename std::decay<F>::type>::value &&
-            !hpx::traits::is_one_way_executor<
-                typename std::decay<F>::type>::value &&
-            !hpx::traits::is_two_way_executor<
-                typename std::decay<F>::type>::value &&
-            !hpx::traits::is_threads_executor<
-                typename std::decay<F>::type>::value
-          , hpx::traits::future_then_result<future, F>
-        >::type
+        typename hpx::traits::future_then_result<future, F>::type
```

Comment:
Something like this perhaps:
```suggestion
        typename util::lazy_enable_if<hpx::traits::is_callable<F(future&&)>::value, typename hpx::traits::future_then_result<future, F>::type>::type
```?

Reply:
And the inverse condition below on `T0`.

Reply:
I'd say rather this
```suggestion
        typename util::lazy_enable_if<hpx::traits::is_callable<F(future&&)>::value, hpx::traits::future_then_result<future, F>>::type
```
You want to delay the instatiation of `future_then_result<>:;type` after all

Reply:
@aurianer mind giving this a try?

Reply:
I did not manage to make it compile, it fails on this :
```
In file included from /users/aurianer/projects/hpx_module_futures/src/runtime_impl.cpp:12:
In file included from /apps/daint/UES/aurianer/build/hpx/build_module_futures/libs/collectives/include/hpx/collectives.hpp:15:
/users/aurianer/projects/hpx_module_futures/libs/collectives/include/hpx/collectives/broadcast.hpp:426:46: error: no matching member
      function for call to 'then'
            hpx::when_all(broadcast_futures).then(&return_void).get();
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
```
Any hints for me ?

Reply:
Ok actually it was my bad sorry ^^

Reply:
Still does not compile for :
```
/users/aurianer/projects/hpx_module_futures/libs/collectives/include/hpx/collectives/gather.hpp:325:27: error: no matching member
      function for call to 'then'
            return result.then(
                   ~~~~~~~^~~~
/users/aurianer/projects/hpx_module_futures/hpx/lcos/future.hpp:957:9: note: candidate template ignored: substitution failure [with F =
      (lambda at /users/aurianer/projects/hpx_module_futures/libs/collectives/include/hpx/collectives/gather.hpp:326:17)]: implicit
      instantiation of undefined template 'hpx::traits::is_callable<(lambda at
      /users/aurianer/projects/hpx_module_futures/libs/collectives/include/hpx/collectives/gather.hpp:326:17), hpx::lcos::future<bool>
      &&>'
        then(F && f, error_code& ec = throws)
        ^
/users/aurianer/projects/hpx_module_futures/hpx/lcos/future.hpp:969:9: note: candidate function template not viable: requires at least 2
      arguments, but 1 was provided
        then(T0 && t0, F && f, error_code& ec = throws)
        ^

```